src/extension_commands.ml

```diff
@@ -131,6 +133,124 @@ let _open_current_dune_file =
   in
   command Extension_consts.Commands.open_current_dune_file handler
 
+let _jump_to_prev_hole, _jump_to_next_hole =
+  let pick_next_hole current_pos ~sorted_non_empty_holes_list:holes =
+    (* we want to find such a range that starts after the current position *)
+    let next_hole =
+      List.find holes ~f:(fun range ->
+          match Position.compare current_pos (Range.start range) with
+          | Ordering.Less -> true
+          | Greater -> false
+          | Equal ->
+            (* we don't want the same range that we're in now; we need the next
+               one *)
+            not (Range.contains range ~positionOrRange:(`Position current_pos)))
+    in
+    (* - if the current position is larger than all other ranges, we cycle back
+       to first hole in the file
+
+       - [hd_exn] because [holes] is non-empty *)
+    Option.value_or next_hole ~default:(fun () -> List.hd_exn holes)
+  in
+
+  let pick_prev_hole current_pos ~sorted_non_empty_holes_list:holes =
+    (* iterate through all hole ranges and pick the one that has start position
+       after the current cursor position *)
+    let rec find_prev_hole prev_range = function
+      | [] -> prev_range
+      | range :: rest -> (
+        if Range.contains range ~positionOrRange:(`Position current_pos) then
+          prev_range
+        else
+          let start = Range.start range in
+          match Position.compare current_pos start with
+          | Ordering.Less -> prev_range
+          | Greater -> find_prev_hole (Some range) rest
+          | Equal ->
+            (* assert false because we check whether the current pos is in this
+               range in the if-expr above *)
+            assert false)
+    in
+    Option.value_or (find_prev_hole None holes) ~default:(fun () ->
+        List.last_exn holes)
```

Comment:
Again, I think that avoiding `last_exn` is both cleaner and avoids extra assumptions:
```ocaml
let rec find_prev_hole prev_range = function
  | [] -> prev_range
  | [hole] -> hole
  ...
```